### PR TITLE
Simple AssetItem functionality

### PR DESCRIPTION
**Pull request by alina612**
Added GetAll, GetOne, Update and Remove functionality for AssetItem. 

The programm.cs file was automatically formatted and I didn't change it itentionally. I only added:

builder.Services.AddScoped<IAssetItemRepository, AssetItemRepository>();
builder.Services.AddScoped<IAssetItemService, AssetItemService>();
builder.Services.AddControllers();

### CODE REVIEW

**Review by devidsdt — APPROVED** [review]

**Review by Sc3l3t0n — CHANGES_REQUESTED** [review]
🔧 Update dtos should normally not have any default values, so you can check for null when updating the entity.

File: src/TuDa.CIMS.Shared/Dtos/UpdateAssetItemDto.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+﻿namespace TuDa.CIMS.Shared.Dtos;
+
+public class UpdateAssetItemDto
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Shared/Dtos/UpdateAssetItemDto.cs:3]: 🤔 This could be made abstract

**Review by Sc3l3t0n — COMMENTED** [review]
⛏️ The dtos can be transformed into a record

File: src/TuDa.CIMS.Api/Controllers/AssetItemController.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    /// <summary>
+    /// Updates an existing AssetItem by its ID using the provided update model.
+    /// If the update is successful, returns a 200 OK response. If an error occurs during the update, an appropriate error response is returned.
+    /// </summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+    /// <param name="updateModel">the model containing the updated values for the AssetItem </param>
+
+    [HttpPut($"{{{nameof(id)}:guid}}/{{{nameof(updateModel)}:AssetItem}}")]
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Controllers/AssetItemController.cs:54]: 🔧 The `AssetItem` should be transferred in the body and not be necessary in the route
```suggestion
    [HttpPut($"{{{nameof(id)}:guid}}")]
```

File: src/TuDa.CIMS.Api/Controllers/AssetItemController.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    public async Task<IActionResult> UpdateAsync(Guid id, UpdateAssetItemDto updateModel)
+    {
+        await _assetItemService.UpdateAsync(id, updateModel);
+        return Ok();
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Controllers/AssetItemController.cs:58]: 🔧 Only send an `Ok` if the service did not fail

File: src/TuDa.CIMS.Api/Controllers/AssetItemController.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    /// If the removal is successful, returns a 200 OK response.
+    /// If an error occurs during the deletion, an appropriate error response is returned.
+    ///</summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+
+    [HttpDelete($"{{{nameof(id)}:guid}}")]
+    public async Task<IActionResult> RemoveAsync(Guid id)
+    {
+        await _assetItemService.RemoveAsync(id);
+        return Ok();
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Controllers/AssetItemController.cs:72]: 🔧 Only send an `Ok` if the service did not fail

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    /// Returns an existing AssetItem with the specific id.
+    /// </summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+    public async Task<AssetItem?> GetOneAsync(Guid id)
+    {
+        return await _context.AssetItems.Where(i => i.Id == id).Include(i => i.Room).SingleAsync();
+    }
+
+    public Task<ErrorOr<Updated>> UpdateAsync(Guid id, AssetItem updateModel) =>
+        throw new NotImplementedException();
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:36]: 🔧 Can be removed safely

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    /// <summary>
+    /// Returns an existing AssetItem with the specific id.
+    /// </summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+    public async Task<AssetItem?> GetOneAsync(Guid id)
+    {
+        return await _context.AssetItems.Where(i => i.Id == id).Include(i => i.Room).SingleAsync();
+    }
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:33]: 🔧 Decide whether to handle errors with `Optionals` like in the return type or with Exceptions, done by the `SingleAsync`

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+        if (existingItem is null)
+        {
+            return Error.NotFound();
+        }
+
+        existingItem.ItemNumber = updateModel.ItemNumber ?? existingItem.Name;
+        existingItem.Note = updateModel.Note ?? existingItem.Name;
+        existingItem.Shop = updateModel.Shop ?? existingItem.Name;
+        existingItem.Name = updateModel.Name ?? existingItem.Name;
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:57]: 🔧 Use the correct field of `existingItem` for updating

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            case (Chemical chemical, UpdateChemicalItemDto update):
+                chemical.Cas = update.Cas ?? chemical.Cas;
+                chemical.Hazards = update.Hazards ?? chemical.Hazards;
+                chemical.Unit = update.Unit ?? chemical.Unit;
+                break;
+            case (Consumable consumable, UpdateConsumableItemDto update):
+                consumable.Manufacturer = update.Manufacturer ?? consumable.Manufacturer;
+                consumable.SerialNumber = update.SerialNumber ?? consumable.SerialNumber;
+                break;
+        }
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:70]: 🤔 Maybe a default branch which returns an Error would be helpful here.

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+        if (existingItem.Room is null)
+        {
+            return Error.NotFound(
+                "Assetitem.update",
+                $"Room id of AssetItem with id {id} not found."
+            );
+        }
+        var room = await _context.Rooms.SingleOrDefaultAsync(r => r.Id == updateModel.RoomId);
+        existingItem.Room.Id = updateModel.RoomId ?? existingItem.Room.Id;
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:80]: 🔧 You should not check if the existing room is null, rather check if the RoomId of the `dto` does not correlate to one in the database and then return an `NotFound`. Also, do not set the ID of the room. Rather, set the Room property to the retrieved room from the database. Before everything, we should check if the room should be updated at all
```suggestion
        if (updateModel.RoomId is not null) {
            var room = await _context.Rooms.SingleOrDefaultAsync(r => r.Id == updateModel.RoomId);
            if (room is null)
            {
                return Error.NotFound(
                    "Assetitem.update",
                    $"Given RoomId {updateModel.RoomId} was not found."
                );
            }
            existingItem.Room = room;
        }
```

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+    ///Removes an AssetItem with the specific id from the database.
+    /// </summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+    public async Task<ErrorOr<Deleted>> RemoveAsync(Guid id)
+    {
+        var itemToRemove = await _context
+            .AssetItems.Include(i => i.Room)
+            .Where(i => i.Id == id)
+            .SingleAsync();
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:94]: 🔧 With the `ErrorOr` as return type, you should use SingleOrDefaultAsync() and check for null to return a `NotFound` if so.

File: src/TuDa.CIMS.Api/Services/AssetItemService.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    /// <param name="id">the unique id of the AssetItem</param>
+    /// <param name="updateModel">the model containing the updated values for the AssetItem </param>
+    /// <returns>  An <see cref="ErrorOr{T}"/> that either contains an error message if an error occurs,
+    /// or the result of the <see cref="UpdateAsync"/> functionality if successful</returns>>
+    public async Task<ErrorOr<Updated>> UpdateAsync(Guid id, UpdateAssetItemDto updateModel)
+    {
+        try
+        {
+            await _assetItemRepository.UpdateAsync(id, updateModel);
+            return Result.Updated;
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Services/AssetItemService.cs:69]: 🔧  Just return the value of the repo

File: src/TuDa.CIMS.Api/Services/AssetItemService.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    /// </summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+    /// <returns>  An <see cref="ErrorOr{T}"/> that either contains an error message if an error occurs,
+    /// or the result of the <see cref="RemoveAsync"/> functionality if successful</returns>>
+    public async Task<ErrorOr<Deleted>> RemoveAsync(Guid id)
+    {
+        try
+        {
+            await _assetItemRepository.RemoveAsync(id);
+            return Result.Deleted;
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Services/AssetItemService.cs:91]: 🔧 Just return the value of service

**Review by Sc3l3t0n — CHANGES_REQUESTED** [review]
I know you were not finished, but I wrote some comments where I saw something

File: src/TuDa.CIMS.Api/Controllers/AssetItemController.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    /// <summary>
+    /// Retrieves all AssetItems from the service and returns them in an appropriate HTTP response.
+    /// </summary>
+    /// <returns> a 200 OK response if the operation is successfully and a 400 BadRequest response if any error occurs </returns>
+    [HttpGet]
+    public async Task<IActionResult> GetAllAsync()
+    {
+        return (await _assetItemService.GetAllAsync())
+            .ToErrorOr()
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Controllers/AssetItemController.cs:28]: 🔧 Need to be removed

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+                chemical.Cas = update.Cas ?? chemical.Cas;
+                chemical.Hazards = update.Hazards ?? chemical.Hazards;
+                chemical.Unit = update.Unit ?? chemical.Unit;
+                break;
+            case (Consumable consumable, UpdateConsumableItemDto update):
+                consumable.Manufacturer = update.Manufacturer ?? consumable.Manufacturer;
+                consumable.SerialNumber = update.SerialNumber ?? consumable.SerialNumber;
+                break;
+            default:
+                return Error.NotFound();
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:68]: ⛏️ Return a Conflict and mention what's wrong here

File: src/TuDa.CIMS.Api/ConfigureServices.cs [hunk 15]
```diff
@@ -11,6 +11,7 @@ public static WebApplication SetupScalar(this WebApplication app)
     {
         app.MapScalarApiReference(options =>
         {
+            options.Servers = [];
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/ConfigureServices.cs:14]: 🌱 Maybe add a comment, that this is needed to clear the cashed IPs to work with Aspire

File: src/TuDa.CIMS.Api/Controllers/AssetItemController.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+            _ => Ok(),
+            err => BadRequest(err)
+        );
+    }
+
+    [HttpGet("test")]
+    public IActionResult Test()
+    {
+        return Ok(JsonSerializer.Serialize<UpdateAssetItemDto>(new UpdateChemicalItemDto()));
+    }
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Controllers/AssetItemController.cs:84]: 🔧 Remove this testing leftover

File: src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+                chemical.Cas = update.Cas ?? chemical.Cas;
+                chemical.Hazards = update.Hazards ?? chemical.Hazards;
+                chemical.Unit = update.Unit ?? chemical.Unit;
+                break;
+            case (Consumable consumable, UpdateConsumableItemDto update):
+                consumable.Manufacturer = update.Manufacturer ?? consumable.Manufacturer;
+                consumable.SerialNumber = update.SerialNumber ?? consumable.SerialNumber;
+                break;
+            default:
+                return Error.NotFound(
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Repositories/AssetItemRepository.cs:71]: 🔧 This should be a conflict or failure

File: src/TuDa.CIMS.Api/Services/AssetItemService.cs [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    /// </summary>
+    /// <param name="id">the unique id of the AssetItem</param>
+
+    public async Task<ErrorOr<AssetItem>> GetOneAsync(Guid id)
+    {
+        try
+        {
+            return (await _assetItemRepository.GetOneAsync(id)) switch
+            {
+                null => Error.Failure(
```
Comment (Sc3l3t0n) [inline src/TuDa.CIMS.Api/Services/AssetItemService.cs:49]: 🔧 Should be `NotFound` as error type

**Review by Sc3l3t0n — CHANGES_REQUESTED** [review]
Some minor things

**Review by Sc3l3t0n — APPROVED** [review]

**Pull request merged**